ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -748,14 +777,33 @@ static bool TryGetVirtualizationRoot(
     int* kauthError)
 {
     PerfSample findRootSample(perfTracer, PrjFSPerfCounter_VnodeOp_GetVirtualizationRoot);
-        
-    *vnodeFsidInode = Vnode_GetFsidAndInode(vnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
-    *root = VirtualizationRoot_FindForVnode(
-        perfTracer,
-        PrjFSPerfCounter_VnodeOp_FindRoot,
-        PrjFSPerfCounter_VnodeOp_FindRoot_Iteration,
-        vnode,
-        context);
+    
+    if (isDelete)
+    {
+        // TODO(Mac): Once #337 is fixed, remove the code that invalidates the cache entry for delete actions.
+        // Currently delete actions invalidate the cache entry to handle the hardlink+delete rename scenario.  If we do not invalidate
+        // the entry we'll find the root for the newly created hardlink and we need to find the root of the path of the file being deleted.
+        // Testing has shown that looking up the root again has consistently yielded the root of the file being deleted.
+        *root = VnodeCache_RefreshRootForVnode(
```

Comment:
I'm with you on the comment here, but I don't see how this matches the code - I may of course be wrong, so let me explain what I *think* this code does. `VnodeCache_RefreshRootForVnode()` internally performs a `VirtualizationRoot_FindForVnode()`, and as we're in the vnode handler, the delete hasn't actually happened, yet so surely when this gets called on a `rename()` we're just getting the original location's root, not the new destination's?

Instead, I think we should be updating the cache to `RootHandle_Indeterminate` from any vnode handler delete events, and effectively treat entries with `RootHandle_Indeterminate` as cache misses.

Reply:
> as we're in the vnode handler, the delete hasn't actually happened, yet so surely when this gets called on a rename() we're just getting the original location's root, not the new destination's?

It's been a while since I've added this code, and so I should really try removing it and remind myself what breaks 😃 

The problem I ran into was not with `rename` but when applications (e.g .NET Core) use hardlink + delete to move/rename a file.

Imagine I have the file `~/Repos/VFS4G/src/a.txt` and I want to rename it to `~/Repos/b.txt` using hardlink + delete.

(From what I remember) the following would occur:

1. `~/Repos/b.txt` hardlink is created
2. vnode cache is updated, and records that that vnode has no virtualization root
3. PreDelete occurs for the delete of `~/Repos/VFS4G/src/a.txt`
4. vnode cache lookup finds the entry added in step 2, and incorrectly concludes that the PreDelete does not need to be delivered to VFS4G.

Without the vnode cache this scenario does work, leading me to believe that *something* was updating the vnode's parent prior to calling the vnode handler for the PreDelete.

It should also be noted that (assuming this problem is specific to hard links) if #337 were complete we probably wouldn't see a functional test failure as `~/Repos/VFS4G/src/a.txt` would've been added to ModifedPaths.dat when the hardlink was created in step 1.

> Instead, I think we should be updating the cache to RootHandle_Indeterminate from any vnode handler delete events, and effectively treat entries with RootHandle_Indeterminate as cache misses.

I like the idea of storing `RootHandle_Indeterminate`.  Perhaps if we were to do this when the hardlink were created (step 1), there would be no need to explicitly call `VnodeCache_RefreshRootForVnode` here.

We might even be able to remove all of the code that passes around `isDelete`.

Thoughts?  

I think the issue of hardlinks and `VnodeCache_RefreshRootForVnode` is the riskier part of this change.  As mentioned above we can potentially remove this code once #337 is complete, but until then I've tried to keep the root lookups working the same way they did before when hardlinks are involved.

/ cc: @jrbriggs - FYI - #337 will help us with correctness, but if hardlinks in/out of the repo end up being popular we'll have to revisit our approach if/when we enable the status cache on Mac.

Reply:
Files with multiple hard links are definitely going to be tricky, and I don't think we could rely on getting appropriate notifications for them without the vnode cache, it's just that the edge cases and "coincidental" correct cases happen to be different.

The main problem with files with multiple hard links is (for us) that there is no way to obtain a reliable list of all the paths to them. I think I suggested this elsewhere, but I think we should  explicitly detect multi-linked files by checking for `va_nlink > 1` and treat them differently, such as broadcasting messages for them to all providers. This only helps if providers keep a (persisted) list of inodes of files they care about together with the path(s) inside "their" repo of course.

Reply:
@pmj thanks again for your thoughts and insights here.  I will file follow up issue(s) as needed for hardlinks, and see how the `RootHandle_Indeterminate` you've suggested works out.

I think if we can keep functional parity with the `RootHandle_Indeterminate` approach we don't need to block this PR on #337.